clippy_lints/src/duration_suboptimal_units.rs

```diff
@@ -0,0 +1,237 @@
+                cx,
+                DURATION_SUBOPTIMAL_UNITS,
+                expr.span,
+                "constructing a `Duration` using a smaller unit when a larger unit would be more readable",
+                |diag| {
+                    let suggestions = vec![
+                        (func_name.ident.span, promoted_constructor.to_string()),
+                        (arg.span, promoted_value.to_string()),
+                    ];
+                    diag.multipart_suggestion_verbose(
+                        format!("try using {promoted_constructor}"),
+                        suggestions,
+                        Applicability::MachineApplicable,
+                    );
+                },
+            );
+        }
+    }
+}
+
+impl DurationSuboptimalUnits {
+    /// Tries to promote the given constructor and value to a bigger time unit and returns the
+    /// promoted constructor name and value.
+    ///
+    /// Returns [`None`] in case no promotion could be done.
+    fn promote(
+        &self,
+        cx: &LateContext<'_>,
+        expr: &'_ Expr<'_>,
+        constructor_name: Symbol,
+        value: u64,
+    ) -> Option<(Symbol, u64)> {
+        let in_const = cx.tcx.hir_is_inside_const_context(expr.hir_id);
+
+        let (best_unit, best_value) = self
+            .units
+            .iter()
+            .skip_while(|unit| unit.constructor_name != constructor_name)
+            .skip(1)
+            .try_fold(
+                (constructor_name, value),
+                |(current_unit, current_value), bigger_unit| {
+                    if let Some(bigger_value) = current_value.div_exact(u64::from(bigger_unit.factor))
+                        && bigger_unit
+                            .stable_since(in_const)
+                            .is_none_or(|v| self.msrv.meets(cx, v))
+                    {
+                        ControlFlow::Continue((bigger_unit.constructor_name, bigger_value))
+                    } else {
+                        // We have to break early, as we can't skip versions, as they are needed to
+                        // correctly calculate the promoted value.
+                        ControlFlow::Break((current_unit, current_value))
+                    }
+                },
+            )
+            .into_value();
+        (best_unit != constructor_name).then_some((best_unit, best_value))
+    }
+}
+
+#[derive(Clone, Copy)]
+struct Unit {
+    /// Name of the constructor on [`Duration`](std::time::Duration) to construct it from the given
+    /// unit, e.g. [`Duration::from_secs`](std::time::Duration::from_secs)
+    constructor_name: Symbol,
+
+    /// The increase factor over the previous (smaller) unit
+    factor: u16,
+
+    /// In what rustc version stable support for this constructor was added.
+    stable_since: Option<RustcVersion>,
+
+    /// In what rustc version stable support for this constructor in const contexts was added.
+    const_stable_since: Option<RustcVersion>,
```

Comment:
Is that really needed? As I wrote in the latest paragraph of [my previous review](https://github.com/rust-lang/rust-clippy/pull/16250#pullrequestreview-3589436617), it looks like you can always "upgrade" units in a `const` context without extra checking.

Removing `const_stable_since` and everything related to it still passes all the tests. Can you produce a test case that would exhibit a failure? If you do, please add it, if not, please remove handling of `const`.